core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterThroughBoolOrAggregation.java

```diff
@@ -0,0 +1,296 @@
+                            .capturedAs(PROJECT)
+                            .with(source().matching(aggregation()
+                                    .matching(PushFilterThroughBoolOrAggregation::isGroupedBoolOrWithMask)
+                                    .capturedAs(AGGREGATION)))));
+        }
+
+        @Override
+        public Pattern<FilterNode> getPattern()
+        {
+            return pattern;
+        }
+
+        @Override
+        public Result apply(FilterNode node, Captures captures, Context context)
+        {
+            return pushFilter(node, captures.get(AGGREGATION), Optional.of(captures.get(PROJECT)), plannerContext, context);
+        }
+    }
+
+    private static Result pushFilter(FilterNode filterNode, AggregationNode aggregationNode, Optional<ProjectNode> projectNode, PlannerContext plannerContext, Context context)
+    {
+        Symbol boolOrSymbol = getOnlyElement(aggregationNode.getAggregations().keySet());
+        Aggregation aggregation = getOnlyElement(aggregationNode.getAggregations().values());
+
+        DomainTranslator.ExtractionResult extractionResult = getExtractionResult(plannerContext, context.getSession(), filterNode.getPredicate());
+        TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+        Expression remainingExpression = extractionResult.getRemainingExpression();
+
+        if (tupleDomain.isNone()) {
+            // Filter predicate is never satisfied. Replace filter with empty values.
+            return Result.ofPlanNode(new ValuesNode(filterNode.getId(), filterNode.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        Optional<Expression> boolSymbolInRemaining = Optional.empty();
+        Optional<Domain> boolSymbolInTupleDomain = Optional.empty();
+
+        if (!remainingExpression.equals(TRUE)) {
+            boolSymbolInRemaining = remainingExpression.children().stream()
+                    .filter(operand -> (operand instanceof Reference && operand.equals(boolOrSymbol.toSymbolReference())))
+                    .map(Expression.class::cast)
+                    .findAny();
+
+            if (boolSymbolInRemaining.isPresent()) {
+                // correlated exists
+                if (!(remainingExpression instanceof Coalesce coalesceExpression)) {
+                    return Result.empty();
+                }
+
+                Optional<Expression> notSupportedOperands = coalesceExpression.children().stream()
+                        .filter(operand -> !(operand instanceof Reference && operand.equals(boolOrSymbol.toSymbolReference())) && !(operand instanceof Constant constant && constant.type().equals(BOOLEAN)))
+                        .map(Expression.class::cast)
+                        .findAny();
+                if (notSupportedOperands.isPresent()) {
+                    return Result.empty();
+                }
+            }
+        }
+
+        if (!tupleDomain.isAll()) {
```

Comment:
I've used `PushFilterThroughCountAggregation#pushFilter` as a reference. Difference is that `Coalesce` does not end up as tupleDomain and boolSymbolDomain for it will be null.